File: ts/kip-do/src/evaluation.ts
/** Host code is bound to exact rule bytes; uploaded artifacts never execute code. */
import { errors } from './errors.js'
import { digest } from './schema/contracts.js'
import {
  evaluateBinaryRule,
  type EvaluationInput,
  type EvaluationRule,
} from './cognitive.js'
import type { Json, JsonMap } from './json.js'

export class EvaluationRules {
  private readonly rules = new Map<string, EvaluationRule>()
  constructor() {
    this.register({ engine: 'kip:binary-stratified-v1' }, (input) => {
      const count = Object.values(input.samples.treatment).reduce(
        (n, items) => n + items.length,
        0,
      )
      if (
        count <
        Math.max(Number(input.trial.quota), input.minimum_independent_attempts)
      )
        return {
          status: 'insufficient',
          effect: null,
          uncertainty: { method: 'hoeffding', alpha: input.parameters.alpha! },
        }
      return evaluateBinaryRule(
        input.rule,
        input.parameters,
        input.trial.comparability as JsonMap,
        input.samples,
      )
    })
  }
  register(artifact: Json, evaluator: EvaluationRule): string {
    const key = digest(artifact)
    if (this.rules.has(key))
      throw errors.constraintViolation(
        'registered rule digest cannot be rebound; publish new rule artifact',
      )
    this.rules.set(key, evaluator)
    return key
  }
  supports(digest: string): boolean {
    return this.rules.has(digest)
  }
  evaluate(input: EvaluationInput): JsonMap {
    const evaluator = this.rules.get(digest(input.rule))
    if (!evaluator)
      throw errors.unsupportedCapability(
        'no trusted host evaluator is registered for pinned rule digest',
      )
    const comparison = evaluator(structuredClone(input))
    digest(comparison)
    return comparison
  }
}

File: rs/anda_cognitive_nexus/src/evaluation.rs
//! Trusted rule bindings. A digest has exactly one evaluator for a Nexus
//! lifetime; loading a new implementation requires a new artifact/version.
use anda_kip::{
    Json, KipError,
    cognitive::{EvaluationInput, EvaluationRule, evaluate_binary_rule},
};
use std::{collections::BTreeMap, sync::Arc};

#[derive(Clone)]
pub struct EvaluationRules(Arc<parking_lot::RwLock<RuleMap>>);
type RuleMap = BTreeMap<String, Arc<dyn EvaluationRule>>;

impl std::fmt::Debug for EvaluationRules {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("EvaluationRules")
            .field("digests", &self.0.read().keys().collect::<Vec<_>>())
            .finish()
    }
}

impl Default for EvaluationRules {
    fn default() -> Self {
        let registry = Self(Arc::new(parking_lot::RwLock::new(BTreeMap::new())));
        registry.register(&serde_json::json!({"engine":"kip:binary-stratified-v1"}),Arc::new(|input:&EvaluationInput| {
            let count:usize=input.samples.treatment.values().map(Vec::len).sum();
            let quota=input.trial["quota"].as_u64().unwrap_or(2).max(input.minimum_independent_attempts);
            if (count as u64)<quota {
                return Ok(serde_json::json!({"status":"insufficient","effect":null,"uncertainty":{"method":"hoeffding","alpha":input.parameters["alpha"]}}));
            }
            evaluate_binary_rule(&input.rule,&input.parameters,&input.trial["comparability"],&input.samples)
        })).expect("builtin rule");
        registry
    }
}

impl EvaluationRules {
    pub fn supports(&self, digest: &str) -> bool {
        self.0.read().contains_key(digest)
    }

    pub fn register(
        &self,
        artifact: &Json,
        rule: Arc<dyn EvaluationRule>,
    ) -> Result<String, KipError> {
        let digest = crate::schema::contracts::digest(artifact)?;
        let mut rules = self.0.write();
        if rules.contains_key(&digest) {
            return Err(KipError::constraint_violation(
                "a registered rule digest cannot be rebound; publish a new rule artifact",
            ));
        }
        rules.insert(digest.clone(), rule);
        Ok(digest)
    }

    pub fn evaluate(&self, input: &EvaluationInput) -> Result<Json, KipError> {
        let digest = crate::schema::contracts::digest(&input.rule)?;
        let rule = self.0.read().get(&digest).cloned().ok_or_else(|| {
            KipError::unsupported_capability(
                "no trusted host evaluator is registered for the pinned rule digest",
            )
        })?;
        let comparison = rule.evaluate(input)?;
        anda_kip::validate_json(&comparison)?;
        Ok(comparison)
    }
}

impl crate::nexus::CognitiveNexus {
    /// Bind trusted host code to exact rule bytes. Restore registrations on
    /// startup; persisted artifacts alone never execute code.
    pub fn register_evaluation_rule(
        &self,
        artifact: &Json,
        rule: Arc<dyn EvaluationRule>,
    ) -> Result<String, KipError> {
        self.store.evaluation_rules.register(artifact, rule)
    }
}
